common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use futures::{
+    task::{Context, Poll},
+    AsyncRead,
+};
+use std::{
+    cmp::min,
+    collections::HashMap,
+    fmt::Debug,
+    hash::Hash,
+    sync::{Arc, Mutex, RwLock},
```

Comment:
use the fancy new `diem_infallible::{Mutex, RwLock}` : p

don't need to unwrap after `.lock()`!